csp/tests/impl/test_pushadapter.py

```diff
@@ -240,6 +240,36 @@ def graph():
         result = list(x[1] for x in result)
         self.assertEqual(result, expected)
 
+    def test_adapter_engine_shutdown(self):
+        class MyPushAdapterImpl(PushInputAdapter):
+            def __init__(self, interval):
+                self._interval = interval
+                self._thread = None
+                self._running = False
+
+            def start(self, starttime, endtime):
+                self._running = True
+                self._thread = threading.Thread(target=self._run)
+                self._thread.start()
+
+            def stop(self):
+                if self._running:
+                    self._running = False
+                    self._thread.join()
+
+            def _run(self):
+                self.engine_shutdown(ValueError("Dummy exception message"))
+
+        MyPushAdapter = py_push_adapter_def("MyPushAdapter", MyPushAdapterImpl, ts[int], interval=timedelta)
+
+        @csp.graph
+        def graph():
+            adapter = MyPushAdapter(timedelta(seconds=1))
+            csp.print("adapter", adapter)
+
+        with self.assertRaisesRegex(Exception, "ValueError: Dummy exception message"):
```

Comment:
See my comment: would be better if we can get it to be a `ValueError` and I believe we can with a little extra work.

Also, you should have some test asserting that the traceback is what we expect.